src/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/TripleDesCngCipher.cs

```diff
@@ -0,0 +1,141 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Diagnostics;
+using System.Security.Cryptography;
+using Internal.NativeCrypto;
+
+namespace Internal.Cryptography
+{
+    internal sealed class TripleDesCngCipher : BasicSymmetricCipher
+    {
+        private static readonly SafeAlgorithmHandle s_hAlgCbc = Open3DesAlgorithm(Cng.BCRYPT_CHAIN_MODE_CBC);
+        private static readonly SafeAlgorithmHandle s_hAlgEcb = Open3DesAlgorithm(Cng.BCRYPT_CHAIN_MODE_ECB);
+
+        private readonly bool _encrypting;
+        private SafeKeyHandle _hKey;
+        private byte[] _currentIv;  // CNG mutates this with the updated IV for the next stage on each Encrypt/Decrypt call.
+                                    // The base IV holds a copy of the original IV for Reset(), until it is cleared by Dispose().
+
+        public TripleDesCngCipher(CipherMode cipherMode, int blockSizeInBytes, byte[] key, byte[] iv, bool encrypting)
+            : base(cipherMode.GetCipherIv(iv), blockSizeInBytes)
+        {
+            _encrypting = encrypting;
+
+            if (IV != null)
+            {
+                _currentIv = new byte[IV.Length];
+            }
+
+            SafeAlgorithmHandle hAlg = GetCipherAlgorithm(cipherMode);
+            _hKey = hAlg.BCryptImportKey(key);
+            Reset();
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            if (disposing)
+            {
+                SafeKeyHandle hKey = _hKey;
+                _hKey = null;
+                if (hKey != null)
+                {
+                    hKey.Dispose();
+                }
+
+                byte[] currentIv = _currentIv;
+                _currentIv = null;
+                if (currentIv != null)
+                {
+                    Array.Clear(currentIv, 0, currentIv.Length);
+                }
+            }
+
+            base.Dispose(disposing);
```

Comment:
No, this cleanup code isn't really shareable with other BasicSymmetricCiphers.  The other BasicSymmetricCipher that I'm adding, the TripleDesOpenSslCipher class, can't use SafeKeyHandle (since that is a Windows-only concept) and doesn't need to have a "currentIv", because OpenSsl's cipher algorithms have their own "Reset" method, so we don't need to keep the original IV around.

On the topic of sharing, I was thinking it might be beneficial to move the [Algorithms.AesImplementation code](https://github.com/dotnet/corefx/blob/master/src/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/AesNativeCryptoTransform.cs) over to this design that I am using here and that is being used in the Sys.Security.Cryptography.Cng project: UniversalCryptoTransform + BasicSymmetricCipher.  There is plenty of duplicated code between these two designs, and I think all the code could converge on the UniversalCryptoTransform classes.  @bartonjs - any thoughts?